### ISSUE TITLE
Help with BINDetect KeyError

### ISSUE THREAD

**Issue by JestlinNg** [loosolab/TOBIAS#286]
Hi am trying to run BINDetect but each time I came up with a KeyError.

I had run ATACorrect and ScoreBigwig as such and it worked completely fine: 
`conda activate tobias
for file in ${OUT}/5_merged/*.pe.q10.sort.rmdup.bam; do 
    base=$(basename "$file" ".pe.q10.sort.rmdup.bam")
    TOBIAS ATACorrect \
        --bam $file  \
        --genome ${OUT}/Reference/GRCh38.p14.genome.fa \
        --peaks ${OUT}/6_Peaks/${base}.combined_peaks.narrowPeak \
        --cores 3 \
        --outdir ${OUT}/9_footprinting/${base}/
done`

`for file in ${OUT}/5_merged/*.pe.q10.sort.rmdup.bam; do 
    base=$(basename "$file" ".pe.q10.sort.rmdup.bam")
    TOBIAS ScoreBigwig \
        --signal /Volumes/LaCie/KU/ATAC/jestlin/9_footprinting/${base}/*_corrected.bw \
        --regions ${OUT}/6_Peaks/${base}.combined_peaks.narrowPeak \
        --cores 3 \
        --output ${OUT}/9_footprinting/${base}/${base}_ATAC_footprints.bw 
done`

However, when I am trying to run BINDetect with the following code I got a Key error. 
`for file in ${OUT}/5_merged/*.pe.q10.sort.rmdup.bam; do 
    base=$(basename "$file" ".pe.q10.sort.rmdup.bam")
    TOBIAS BINDetect \
        --motifs "${OUT}/Reference/JASPAR/JASPAR2024_CORE_vertebrates_non-redundant_pfms_jaspar.txt" \
        --signals "${OUT}/9_footprinting/${base}/${base}_ATAC_footprints.bw"  \
        --genome "${OUT}/Reference/GRCh38.p14.genome.fa" \
        --peaks "${OUT}/6_Peaks/${base}.combined_peaks.narrowPeak" \
        --cores 8 \
        --outdir "${OUT}/9_footprinting/BINDetect/${base}"
done`

error: 

> ----- Input parameters -----
>signals:      ['CTL_ATAC_footprints.bw']
>peaks:        /Volumes/LaCie/KU/ATAC/6_Peaks/CTL.combined_peaks.narrowPeak
>motifs:       ['/Volumes/LaCie/KU/ATAC/Reference/JASPAR/JASPAR2024_CORE_vertebrates_non-redundant_pfms_jaspar.txt']
>genome:       /Volumes/LaCie/KU/ATAC/Reference/GRCh38.p14.genome.fa
>cond_names:   ['CTL_ATAC_footprints']
>peak_header:  None
>motif_pvalue: 0.0001
>bound_pvalue: 0.001
>pseudo:       None
>time_series:  False
>skip_excel:   False
>prefix:       bindetect
>outdir:       /Volumes/LaCie/KU/ATAC/9_footprinting/CTL
>cores:        8
>split:        100
>verbosity:    3
>
>
>----- Output files -----
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/*/beds/*_CTL_ATAC_footprints_bound.bed
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/*/beds/*_CTL_ATAC_footprints_unbound.bed >/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/*/beds/*_all.bed
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/*/plots/*_log2fcs.pdf
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/*/*_overview.txt
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/*/*_overview.xlsx
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/bindetect_distances.txt
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/bindetect_results.txt
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/bindetect_results.xlsx
>/Volumes/LaCie/KU/ATAC/9_footprinting/CTL/bindetect_figures.pdf
>
>
>2024-08-14 09:41:03 (4290) [INFO]       ----- Processing input data -----
>2024-08-14 09:41:03 (4290) [INFO]       Checking reading/writing of files
>2024-08-14 09:41:05 (4290) [INFO]       - Found 178744 regions in input peaks
>2024-08-14 09:41:05 (4290) [INFO]       - Merged to 136432 regions
>2024-08-14 09:41:06 (4290) [INFO]       Estimating GC content from peak sequences
>2024-08-14 09:41:29 (4290) [INFO]       - GC content estimated at 50.03%
>2024-08-14 09:41:29 (4290) [INFO]       Reading motifs from file
>2024-08-14 09:41:29 (4290) [INFO]       - Read 879 motifs
>2024-08-14 09:41:31 (4290) [INFO]       Creating folder structure for each TF
>
>2024-08-14 09:42:27 (4290) [INFO]       Scanning for motifs and matching to signals...
>2024-08-14 09:42:27 (4290) [INFO]       Progress 0%
>ERROR IN REGION: KI270466.1     839     1248    CTL.combined_peak_54    357     .       11.41614        38.17934        >35.78844        200
>2024-08-14 09:42:29 (4344) [ERROR]      Error reading footprints from region: KI270466.1        839     1248    >CTL.combined_peak_54    357     .       11.41614        38.17934   35.78844 200
>2024-08-14 09:42:29 (4290) [INFO]       Progress 1.0%
>2024-08-14 09:42:38 (4290) [INFO]       Progress 2.0%

> 2024-08-14 09:42:38 (4290) [INFO]       Progress 3.0%
> 2024-08-14 09:42:39 (4290) [INFO]       Progress 4.0%
> 2024-08-14 09:42:39 (4290) [INFO]       Progress 5.0%
> 2024-08-14 09:42:39 (4290) [INFO]       Progress 7.0%
> 2024-08-14 09:42:40 (4290) [INFO]       Progress 8.0%
> 2024-08-14 09:42:48 (4290) [INFO]       Progress 9.0%
> 2024-08-14 09:42:49 (4290) [INFO]       Progress 10.0%
> 2024-08-14 09:42:49 (4290) [INFO]       Progress 11.0%
> 2024-08-14 09:42:49 (4290) [INFO]       Progress 12.0%
> 2024-08-14 09:42:50 (4290) [INFO]       Progress 13.0%
> 2024-08-14 09:42:50 (4290) [INFO]       Progress 14.0%
> 2024-08-14 09:42:50 (4290) [INFO]       Progress 15.0%
> 2024-08-14 09:42:59 (4290) [INFO]       Progress 16.0%
> 2024-08-14 09:42:59 (4290) [INFO]       Progress 17.0%
> 2024-08-14 09:43:00 (4290) [INFO]       Progress 19.0%
> 2024-08-14 09:43:00 (4290) [INFO]       Progress 21.0%
> 2024-08-14 09:43:01 (4290) [INFO]       Progress 22.0%
> 2024-08-14 09:43:10 (4290) [INFO]       Progress 23.0%
> 2024-08-14 09:43:10 (4290) [INFO]       Progress 25.0%
> 2024-08-14 09:43:10 (4290) [INFO]       Progress 26.0%
> 2024-08-14 09:43:11 (4290) [INFO]       Progress 27.0%
> 2024-08-14 09:43:11 (4290) [INFO]       Progress 28.0%
> 2024-08-14 09:43:11 (4290) [INFO]       Progress 29.0%
> 2024-08-14 09:43:21 (4290) [INFO]       Progress 30.0%
> 2024-08-14 09:43:21 (4290) [INFO]       Progress 31.0%
> 2024-08-14 09:43:21 (4290) [INFO]       Progress 32.0%
> 2024-08-14 09:43:22 (4290) [INFO]       Progress 34.0%
> 2024-08-14 09:43:22 (4290) [INFO]       Progress 35.0%
> 2024-08-14 09:43:22 (4290) [INFO]       Progress 36.0%
> 2024-08-14 09:43:31 (4290) [INFO]       Progress 37.0%
> 2024-08-14 09:43:32 (4290) [INFO]       Progress 38.0%
> 2024-08-14 09:43:32 (4290) [INFO]       Progress 39.0%
> 2024-08-14 09:43:32 (4290) [INFO]       Progress 41.0%
> 2024-08-14 09:43:33 (4290) [INFO]       Progress 42.0%
> 2024-08-14 09:43:33 (4290) [INFO]       Progress 43.0%
> 2024-08-14 09:43:42 (4290) [INFO]       Progress 44.0%
> 2024-08-14 09:43:43 (4290) [INFO]       Progress 46.0%
> 2024-08-14 09:43:43 (4290) [INFO]       Progress 47.0%
> 2024-08-14 09:43:43 (4290) [INFO]       Progress 48.0%
> 2024-08-14 09:43:44 (4290) [INFO]       Progress 50.0%
> 2024-08-14 09:43:53 (4290) [INFO]       Progress 51.0%
> 2024-08-14 09:43:53 (4290) [INFO]       Progress 53.0%
> 2024-08-14 09:43:54 (4290) [INFO]       Progress 55.0%
> 2024-08-14 09:43:54 (4290) [INFO]       Progress 56.0%
> 2024-08-14 09:43:54 (4290) [INFO]       Progress 57.0%
> 2024-08-14 09:44:03 (4290) [INFO]       Progress 58.0%
> 2024-08-14 09:44:04 (4290) [INFO]       Progress 59.0%
> 2024-08-14 09:44:04 (4290) [INFO]       Progress 60.0%
> 2024-08-14 09:44:04 (4290) [INFO]       Progress 61.0%
> 2024-08-14 09:44:04 (4290) [INFO]       Progress 62.0%
> 2024-08-14 09:44:05 (4290) [INFO]       Progress 63.0%
> 2024-08-14 09:44:05 (4290) [INFO]       Progress 64.0%
> 2024-08-14 09:44:14 (4290) [INFO]       Progress 65.0%
> 2024-08-14 09:44:14 (4290) [INFO]       Progress 66.0%
> 2024-08-14 09:44:15 (4290) [INFO]       Progress 67.0%
> 2024-08-14 09:44:15 (4290) [INFO]       Progress 68.0%
> 2024-08-14 09:44:15 (4290) [INFO]       Progress 69.0%
> 2024-08-14 09:44:15 (4290) [INFO]       Progress 70.0%
> 2024-08-14 09:44:16 (4290) [INFO]       Progress 71.0%
> 2024-08-14 09:44:24 (4290) [INFO]       Progress 72.0%
> 2024-08-14 09:44:24 (4290) [INFO]       Progress 73.0%
> 2024-08-14 09:44:25 (4290) [INFO]       Progress 74.0%
> 2024-08-14 09:44:25 (4290) [INFO]       Progress 75.0%
> 2024-08-14 09:44:25 (4290) [INFO]       Progress 76.0%
> 2024-08-14 09:44:26 (4290) [INFO]       Progress 78.0%
> 2024-08-14 09:44:35 (4290) [INFO]       Progress 80.0%
> 2024-08-14 09:44:36 (4290) [INFO]       Progress 81.0%
> 2024-08-14 09:44:36 (4290) [INFO]       Progress 82.0%
> 2024-08-14 09:44:37 (4290) [INFO]       Progress 83.0%
> 2024-08-14 09:44:37 (4290) [INFO]       Progress 84.0%
> 2024-08-14 09:44:37 (4290) [INFO]       Progress 85.0%
> 2024-08-14 09:44:46 (4290) [INFO]       Progress 86.0%
> 2024-08-14 09:44:47 (4290) [INFO]       Progress 87.0%
> 2024-08-14 09:44:47 (4290) [INFO]       Progress 88.0%
> 2024-08-14 09:44:47 (4290) [INFO]       Progress 89.0%
> 2024-08-14 09:44:48 (4290) [INFO]       Progress 90.0%
> 2024-08-14 09:44:48 (4290) [INFO]       Progress 91.0%
> 2024-08-14 09:44:48 (4290) [INFO]       Progress 92.0%
> 2024-08-14 09:44:57 (4290) [INFO]       Progress 93.0%
> 2024-08-14 09:44:57 (4290) [INFO]       Progress 94.0%
> 2024-08-14 09:44:57 (4290) [INFO]       Progress 95.0%
> 2024-08-14 09:44:58 (4290) [INFO]       Progress 96.0%
> 2024-08-14 09:44:58 (4290) [INFO]       Progress 97.0%
> 2024-08-14 09:44:58 (4290) [INFO]       Progress 99.0%
> 2024-08-14 09:45:06 (4290) [INFO]       Progress 100.0%
> 2024-08-14 09:45:06 (4290) [INFO]       Progress done!
> multiprocessing.pool.RemoteTraceback: 
> """
> Traceback (most recent call last):
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/multiprocessing/pool.py", line 121, in worker
>     result = (True, func(*args, **kwds))
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/site-packages/tobias/footprinting/bindetect_functions.py", line 234, in scan_and_score
>     bigwig_score = footprints[bigwig][pos]
> KeyError: 'CTL_ATAC_footprints'
> """
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/opt/anaconda3/envs/tobias/bin/TOBIAS", line 11, in <module>
>     load_entry_point('tobias==0.8.0', 'console_scripts', 'TOBIAS')()
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/site-packages/tobias/TOBIAS.py", line 152, in main
>     args.func(args)             #run specified function with arguments
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/site-packages/tobias/footprinting/bindetect.py", line 332, in run_bindetect
>     results = [task.get() for task in task_list]
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/site-packages/tobias/footprinting/bindetect.py", line 332, in <listcomp>
>     results = [task.get() for task in task_list]
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/multiprocessing/pool.py", line 657, in get
>     raise self._value
> KeyError: 'CTL_ATAC_footprints'
> Problem in main logger process:
> Traceback (most recent call last):
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/site-packages/tobias/utils/logger.py", line 147, in main_logger_process
>     record = self.queue.get()
>   File "<string>", line 2, in get
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/multiprocessing/managers.py", line 819, in _callmethod
>     kind, result = conn.recv()
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/multiprocessing/connection.py", line 250, in recv
>     buf = self._recv_bytes()
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/multiprocessing/connection.py", line 407, in _recv_bytes
>     buf = self._recv(4)
>   File "/opt/anaconda3/envs/tobias/lib/python3.7/multiprocessing/connection.py", line 383, in _recv
>     raise EOFError
> EOFError

**Comment by mohobein** [comment]
Hey @JestlinNg,

looks like there is a problem with the input file, an end of file error specifically that does not allow BINDetect to process the scores correctly. To investigate further, could you please send me the last few lines of your signal and peak files, or just the whole files if possible? Then I can try to reproduce it and see if I can spot the problem and fix it.

Kind regards,
Moritz

**Comment by JestlinNg** [comment]
Attachments available until Sep 21, 2024
Hi Moritz, thank you for offering to help me check it out. Here are my narrowPeak and signal files.
Click to Download<https://www.icloud.com/attachment/?u=https%3A%2F%2Fcvws.icloud-content.com%2FB%2FAX-IvggpHYvnb6NY3rWuo590RxMrAWP_0HdWrwQaFmD0hTvjdOJ4jVWI%2F%24%7Bf%7D%3Fo%3DApbrFWOv6uKOctTA0jHfsFUaICkenPhbZrny7VtNI-rx%26v%3D1%26x%3D3%26a%3DCAog13vGK8VcDG7ysxMTgwskvmoXk77v1PJwuYhCZDHGsEcSdhCT1ZbElzIYk-WRmKEyIgEAKgkC6AMA_xhysrlSBHRHEytaBHiNVYhqJedkB_fTV6VU0v4F7TScYSit7gI1G-Z9j3-a690UqDeMAW9gDDVyJcyGYXGKfeZuRF5K3ISADJfyzhT4qbgLfPwVL0-iqxY9iwdW1fM%26e%3D1726895911%26fl%3D%26r%3D56AF1007-20CF-4698-BF50-3AD810B13958-1%26k%3D%24%7Buk%7D%26ckc%3Dcom.apple.largeattachment%26ckz%3D42C01E99-158F-4196-9972-956FD0FC8626%26p%3D110%26s%3DlRjYis6wFKVzvbhDpUupoIZ7HOE&uk=xUKNg2VD2pyGOSF0nXlv2w&f=CTL.combined_peaks.narrowPeak&sz=15003904>
CTL.combined_peaks.narrowPeak
15 MB
Click to Download<https://www.icloud.com/attachment/?u=https%3A%2F%2Fcvws.icloud-content.com%2FB%2FAYs6VUfY3d6qXpawGbpPN4tNX-itAaph2C92O5gDWL2EQ9hD25CykcEs%2F%24%7Bf%7D%3Fo%3DAkJWuKAsd_Vsxpz6oBAok3aRs_VmD7nMltB7_K9elZfq%26v%3D1%26x%3D3%26a%3DCAogEgojYyTU2qxaBeV7WlWz0iO4HPlzG1aY7WyaKXy0g4kSdhC02ZbElzIYtOmRmKEyIgEAKgkC6AMA_wzmLx1SBE1f6K1aBLKRwSxqJXMotPm3bvw0jo6XG9xodiaGnuowf2Ny6Dd8DSHrvT8axvxTcThyJfcGDDm1W1jgTDCpZKHasa4EH0pSC-yzoxxa5AxWKVuYyzR_Pe0%26e%3D1726895912%26fl%3D%26r%3DC3E6001D-A8A9-4828-9D6D-DE08968B9E17-1%26k%3D%24%7Buk%7D%26ckc%3Dcom.apple.largeattachment%26ckz%3D42C01E99-158F-4196-9972-956FD0FC8626%26p%3D110%26s%3D1UIUOkUUwQuqDKGb4mUpPVdGO0M&uk=f0xEV-VZzFrvgPHJ2QS8lA&f=CTL_ATAC_footprints.bw&sz=420964036>
CTL_ATAC_footprints.bw
421 MB


On Aug 19, 2024, at 11:28 PM, Moritz Hobein ***@***.******@***.***>> wrote:



Hey @JestlinNg<https://github.com/JestlinNg>,

looks like there is a problem with the input file, an end of file error specifically that does not allow BINDetect to process the scores correctly. To investigate further, could you please send me the last few lines of your signal and peak files, or just the whole files if possible? Then I can try to reproduce it and see if I can spot the problem and fix it.

Kind regards,
Moritz

**Comment by mohobein** [comment]
Hey,

I tried to replicate your error. I think the issue might be related to the fact that your peak file contains regions outside of the standard chromosomes. When I filtered out those regions that were not placed on the main chromosomes (by creating a new peak file using `grep "chr" CTL.combined_peaks.narrowPeak > filtered_peaks.bed`), the tool had no issues. Are these regions important to your analysis? If not, it might be easiest to just exclude them. Otherwise, I'll have to investigate further to see what exactly causes this problem.

Best regards,
Moritz

**Comment by JestlinNg** [comment]
Oh!! I see! Those regions aren't important so I will remove them accordingly and run it again. Thank you so so much for the help ~

On 30 Aug 2024, at 9:12 p.m., Moritz Hobein ***@***.***> wrote:

﻿

Hey,

I tried to replicate your error. I think the issue might be related to the fact that your peak file contains regions outside of the standard chromosomes. When I filtered out those regions that were not placed on the main chromosomes (by creating a new peak file using grep "chr" CTL.combined_peaks.narrowPeak > filtered_peaks.bed), the tool had no issues. Are these regions important to your analysis? If not, it might be easiest to just exclude them. Otherwise, I'll have to investigate further to see what exactly causes this problem.

Best regards,
Moritz
